8330144: Revise os::free_memory()

roberttoyonaga opened the pull request:
### Summary
On linux, change `os::free_memory(char *addr, size_t bytes, size_t alignment_hint)` so that it uses `madvise(MADV_DONTNEED)` (similar to the BSD implementation) instead of recommitting over the existing committed memory to discard the existing pages. This function should free the underlying memory without uncommitting.  The benefit of this change is that we can get rid of conditional logic dependent on whether we're dealing with huge pages, `madvise` can't fail, and we can also get rid of the "alignment_hint" parameter. 

`os::free_memory(char *addr, size_t bytes, size_t alignment_hint)` has also been renamed to `os::disclaim_memory(char *addr, size_t bytes)` to differentiate it from `os::free_memory()` which reports the size of free memory instead of actually releasing memory.  

**Transparent huge pages:**
`madvise(MADV_DONTNEED)` works with THP. As with small pages, `madvise(MADV_DONTNEED)` results in the memory being freed, RSS decreasing, and the addresses can be re-touched without being explicitly recommitted.

To determine this, I set /sys/kernel/mm/transparent_hugepage/enabled to "always" and allocated a large amount of memory. Then /proc/PID/smaps shows that THP are being used to back that memory. After calling `disclaim_memory`, RSS decreases indicating the memory is no longer live. The `os::committed_in_range function` also reports that the memory has been freed (This function should probably be renamed to `live_in_range`). Touching the addresses again afterward is fine as well. 

**Explicit huge pages:**
`madvise(MADV_DONTNEED)` does not result in memory being freed when used on explicit huge pages. However, the pages are not lost either. Additionally, after `madvise(MADV_DONTNEED)`, we can retouch the addresses without any problems. In conclusion, `madvise(MADV_DONTNEED)` has no affect on huge pages. This means the behavior of of this function with respect to huge pages remains the same. We can remove the "alignment_hint" parameter.

To determine this, I allocated some huge pages via /proc/sys/vm/nr_hugepages. Successful allocation was confirmed with /proc/meminfo.  After calling `disclaim_memory`, /proc/meminfo shows no change in the number of huge pages in use.  Explicit huge pages are not reflected in RSS so I used the `os::committed_in_range function` instead.  After calling `disclaim_memory`, the `os::committed_in_range` function reports that the memory is still live. Unfortunately that's not an improvement upon existing behavior, but at least it's not a regression either.

#### Testing

- Added the gtest: free_without_uncommit. This test is excluded for AIX and Windows since on those platforms `disclaim_memory` does nothing. Interestingly, unlike linux, [madvise(DONTNEED) on BSD](https://man.freebsd.org/cgi/man.cgi?query=madvise&sektion=2&n=1) doesn't free pages, it only lowers their priority, increasing likelihood of future page faults. So this test has special handling is for BSD.  If our intention is to have the pages freed, maybe MADV_FREE is a better choice.
- tier1

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8330144](https://bugs.openjdk.org/browse/JDK-8330144): Revise os::free_memory() (**Enhancement** - P4)


### Reviewers
 * [Thomas Stuefe](https://openjdk.org/census#stuefe) (@tstuefe - **Reviewer**)
 * [Matthias Baesken](https://openjdk.org/census#mbaesken) (@MBaesken - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/20080/head:pull/20080` \
`$ git checkout pull/20080`

Update a local copy of the PR: \
`$ git checkout pull/20080` \
`$ git pull https://git.openjdk.org/jdk.git pull/20080/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 20080`

View PR using the GUI difftool: \
`$ git pr show -t 20080`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/20080.diff">https://git.openjdk.org/jdk/pull/20080.diff</a>

</details>


### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/20080#issuecomment-2218110653)

roberttoyonaga commented:
Failing GHA:

-[linux-x86 / test (hs/tier1 compiler part 2)](https://github.com/roberttoyonaga/jdk/actions/runs/9857950912/job/27219620860): compiler/interpreter/Test6833129.java encounters an error with exit code 1. I don't think this is related, since these changes don't relate to compilation.

tstuefe commented on test/hotspot/gtest/runtime/test_os.cpp:
```diff
@@ -976,3 +979,34 @@ TEST_VM(os, vm_min_address) {
+  const size_t pages = 64;
+  const size_t size = pages * page_sz;
+
+  char *base = os::reserve_memory(size, false, mtTest);
```
I prefer char* base (star at type) syntax, and its much more common in hotspot.

tstuefe commented on test/hotspot/gtest/runtime/test_os.cpp:
```diff
@@ -976,3 +979,34 @@ TEST_VM(os, vm_min_address) {
+  // Check that memory has been freed. Skip on BSD since MADV_DONTNEED doesn't free memory.
+  size_t committed_size;
+  address committed_start;
+  ASSERT_FALSE(os::committed_in_range((address) base, size, committed_start, committed_size));
```
Is there a chance of this generating false positives? Do we know if the madvise effect immediate or delayed?

roberttoyonaga replied:
That's a good point. Based on the linux [docs](https://man7.org/linux/man-pages/man2/madvise.2.html) it might not happen immediately, causing the test to fail. I'll remove the `committed_in_range check`.  I suppose we could poll with a timeout, but there's still no guarantee the pages actually get freed in a timely manner.

tstuefe reviewed (changes requested):
Great, thanks @roberttoyonaga. The main work was the analysis work beforehand.

About naming, I would name the thing "os::disclaim_memory". free_without_uncommit is a mouthful. There is a precedence in the "disclaim" API on AIX, which in a future RFE may be used to implement os::disclaim_memory.

roberttoyonaga commented:
Thank you @tstuefe for the review feedback! I've renamed to `free_memory_without_uncommit` to `disclaim_memory` and removed the `committed_in_range` check so the unit test can be more reliable.

tstuefe reviewed (changes requested):
minor nits, fine otherwise

roberttoyonaga commented on src/hotspot/os/windows/os_windows.cpp:
```diff
@@ -3893,7 +3893,7 @@ bool os::unguard_memory(char* addr, size_t bytes) {
 
 void os::pd_realign_memory(char *addr, size_t bytes, size_t alignment_hint) { }
-void os::pd_free_memory(char *addr, size_t bytes, size_t alignment_hint) { }
+void os::pd_disclaim_memory(char *addr, size_t bytes) { }
```
Ok I've added a comment with a description. 

Is it good practice to add these types of descriptions in the shared code header files (os.hpp), in the platform dependent code (os_linux.hpp), or both? I see some examples of all 3 cases, but I'm wondering if there's a best practice.

tstuefe replied:
There is no common format. Sun did not comment any APIs in the beginning. I usually do it above the prototype in the hpp file, because then IDEs cab pick it up and show you the description in tooltips.

roberttoyonaga commented:
/integrate

Thank you @tstuefe for the review!

tstuefe commented:
Hi @roberttoyonaga, unfortunately you'll need a second reviewer (standard rule for hotspot changes).

@MBaesken maybe?

MBaesken commented:
The comments above and coding looks okay to me.
Will test it first in our CI .
Are you sure there are no side effects or maybe 'bad' kernel versions where unwanted things occur when using madvice instead?

tstuefe commented:
> The comments above and coding looks okay to me. Will test it first in our CI . Are you sure there are no side effects or maybe 'bad' kernel versions where unwanted things occur when using madvice instead?

@MBaesken Reasonably sure. The same technique has been used by the glibc for C-heap trimming for a long time. Thanks for putting this into SAP's CI!

tstuefe commented:
/sponsor
